proposals/readonly-struct-init.md

```diff
@@ -0,0 +1,161 @@
+# readonly object initializers 
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Below
+
+## Summary
+
+This feature will allow for assigning of `readonly` fields and get only auto-implemented properties
+in object initializers when the type being constructed is a `struct` or `ref struct`.
+
+## Motivation
```

Comment:
I'm concerned about this feature breaking encapsulation. If the constructor has logic that safeguards the state of the object, then assigning to a field or property directly could violate it.
Note there is another proposal which is similar: ["data" types](https://github.com/dotnet/csharplang/pull/1667), which *requires* fields or properties to be initializable from a constructor initializer. But the choice is explicit (`data` modifier in type declaration).

If we move forward with the `data` types proposal, would you still push the present proposal forward?

Reply:
>  If the constructor has logic that safeguards the state of the object, then assigning to a field or property directly could violate it.

How is this different than constructor logic that safeguards mutable fields?

Reply:
A constructor does not and cannot safeguard a (public) mutable field, just like a sieve does not and cannot hold water.

Currently, if you want to protect invariants with encapsulation, you make the mutable parts inaccessible or you put additional logic in the setters. But the present proposal opens up an accessible and unguarded setter. That worries me.

Reply:
The contract of mutability. Constructor logic on mutable fields can be broken, and if you want that logic generally applicable you need to also place it in the property setter (and should not expose the field).

While `readonly` properties/fields today are believed to only be assignable from the constructor, so the logic only needs to exist there (such as saying that an integer must be positive).

Reply:
I'm really not too worried about publicly exposed fields being mutated. Structs don't own their storage and hence don't have the same guarantees as classes do for their state. It's easy to mutate struct contents today without this feature.

Reply:
The same is true for classes. Mutating struct state to violate its internal invariant is about as hard as mutating class state to violate its internal invariant.